ydb/core/kqp/provider/rewrite_io_utils.cpp

```diff
@@ -0,0 +1,169 @@
+    AddChild(carrier, ctx.NewCallable(payload->Pos(), QueryGraphNodeSignature, {payload}));
+}
+
+void InsertExecutionOrderDependencies(const TExprNode::TPtr& queryGraph, const TExprNode::TPtr& worldBefore) {
+    VisitExpr(
+        queryGraph,
+        nullptr,
+        [&worldBefore](const TExprNode::TPtr& node) {
+            if (node->ChildrenSize() > 0 && node->Child(0)->IsWorld()) {
+                ChangeChild(node, 0, worldBefore);
+            }
+            return true;
+        }
+    );
+}
+
+bool CheckTopLevelness(const TExprNode::TPtr& candidateRead, const TExprNode::TPtr& queryGraph) {
+    THashSet<TExprNode::TPtr> readsInCandidateSubgraph;
+    VisitExpr(candidateRead, [&readsInCandidateSubgraph](const TExprNode::TPtr& node) {
+        if (node->IsCallable(ReadName)) {
+            readsInCandidateSubgraph.emplace(node);
+        }
+        return true;
+    });
+
+    return !FindNode(queryGraph, [&readsInCandidateSubgraph](const TExprNode::TPtr& node) {
+        return node->IsCallable(ReadName) && !readsInCandidateSubgraph.contains(node);
+    });
+}
+
+TExprNode::TPtr FindTopLevelRead(const TExprNode::TPtr& queryGraph) {
+    const TExprNode::TPtr* lastReadInTopologicalOrder = nullptr;
+    VisitExpr(
+        queryGraph,
+        nullptr,
+        [&lastReadInTopologicalOrder](const TExprNode::TPtr& node) {
+            if (node->IsCallable(ReadName)) {
+                lastReadInTopologicalOrder = &node;
+            }
+            return true;
+        }
+    );
+
+    if (!lastReadInTopologicalOrder) {
+        return nullptr;
+    }
+
+    YQL_ENSURE(CheckTopLevelness(*lastReadInTopologicalOrder, queryGraph),
+               "Info for developers: assumption that there is only one top level Read! is wrong "
+               "for the expression graph of the query stored in the view:\n"
+                   << queryGraph->Dump());
+
+    return *lastReadInTopologicalOrder;
+}
+
+}
+
+TExprNode::TPtr RewriteReadFromView(
+    const TExprNode::TPtr& node,
+    TExprContext& ctx,
+    const TString& query,
+    const TString& cluster
+) {
+    const auto readNode = node->ChildPtr(0);
+    const auto worldBeforeThisRead = readNode->ChildPtr(0);
+
+    TExprNode::TPtr queryGraph = FindSavedQueryGraph(readNode);
+    if (!queryGraph) {
+        queryGraph = CompileQuery(query, ctx, cluster);
+        if (!queryGraph || ContainsNullNode(queryGraph)) {
+            ctx.AddError(TIssue(readNode->Pos(ctx),
+                         "The query stored in the view contains errors and cannot be compiled."));
+            return nullptr;
+        }
+        YQL_CLOG(TRACE, ProviderKqp) << "Expression graph of the query stored in the view:\n"
+                                     << NCommon::ExprToPrettyString(ctx, *queryGraph);
+
+        InsertExecutionOrderDependencies(queryGraph, worldBeforeThisRead);
+        SaveQueryGraph(readNode, ctx, queryGraph);
```

Comment:
I'm not quite sure why do we need to save view AST to Read? Why not just rewrite the query immediately?

Reply:
I explained the motivation in [this comment](https://github.com/ydb-platform/ydb/pull/795#discussion_r1438257676). The idea is that in most cases we use nodes that express Read!s from views in two ways: 
1) call Left! on the read node,
2) call Right! on that same read node.

I noticed that in order for all the nodes with side-effects (like Read! from tables and such) in the rewritten query graph of the select from the view statement to be linearly stacked on top of each other, we need to ensure that the query graph of the view query is exactly the same in both rewrites: a) of the Left! call and b) of the Right! call to the node reading from the view. Otherwise, the rewritten query will fail on CheckTx function.

I didn't find a better way to ensure that we call CompileExpr on the view query only once.

Reply:
ок